pkg/kubelet/kuberuntime/kuberuntime_gc.go

```diff
@@ -0,0 +1,231 @@
+	if n <= 0 {
+		return
+	}
+
+	toKeep := len(gc.containers) - n
+	for i := toKeep; i < len(gc.containers); i++ {
+		gc.removeContainer(gc.containers[i].containerID)
+	}
+
+	gc.containers = gc.containers[:toKeep]
+}
+
+// Newest first.
+type containerByCreated []*containerGCInfo
+
+func (a containerByCreated) Len() int           { return len(a) }
+func (a containerByCreated) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a containerByCreated) Less(i, j int) bool { return a[i].createdAt.After(a[j].createdAt) }
+
+// toContainerGCInfo converts kubecontainer.ContainerStatus to containerGCInfo.
+func toContainerGCInfo(podUID, sandboxID string, c *kubecontainer.ContainerStatus) *containerGCInfo {
+	return &containerGCInfo{
+		containerID: c.ID.ID,
+		createdAt:   c.CreatedAt,
+		sandboxID:   sandboxID,
+		podUID:      podUID,
+	}
+}
+
+// getEvictablePods gets evictable pods.
+// Evictable pods include sandboxes and containers which are not active and created more than MinAge ago).
+func (m *kubeGenericRuntimeManager) getEvictablePods(minAge time.Duration) (map[string]*podGCInfo, error) {
+	sandboxes, err := m.getKubeletSandboxes(true)
+	if err != nil {
+		return nil, err
+	}
+
+	evictablePods := make(map[string]*podGCInfo)
+	newestGCTime := time.Now().Add(-minAge)
+	for _, s := range sandboxes {
+		if s.GetState() != runtimeApi.PodSandBoxState_NOTREADY {
+			continue
+		}
+
+		createdAt := time.Unix(*s.CreatedAt, 0)
+		if newestGCTime.Before(createdAt) {
+			continue
+		}
+
+		podUID := getStringValueFromLabel(s.Labels, types.KubernetesPodUIDLabel)
+		if len(podUID) == 0 {
+			podUID = nonKubernetesPodUID
+		}
+		if pod, ok := evictablePods[podUID]; ok {
+			pod.sandboxIDs = append(pod.sandboxIDs, s.GetId())
+		} else {
+			evictablePods[podUID] = &podGCInfo{
+				client:     m.runtimeService,
+				containers: make([]*containerGCInfo, 0),
+				sandboxIDs: []string{s.GetId()},
+			}
+		}
+
+		containerStatuses, err := m.getKubeletContainerStatuses(s.GetId())
+		if err != nil {
+			return nil, err
+		}
+		for _, c := range containerStatuses {
+			evictablePods[podUID].containers = append(evictablePods[podUID].containers,
+				toContainerGCInfo(podUID, s.GetId(), c))
+		}
+
+		sort.Sort(containerByCreated(evictablePods[podUID].containers))
+	}
+
+	return evictablePods, nil
+}
+
+// GarbageCollect removes dead containers using the specified container gc policy.
+// It :
+// * gets evictable pods with belonging containers and sandboxes which are not active and created more than MinAge ago.
+// * remove oldest dead containers for each Pod by enforcing MaxPerPodContainer
+// * remove oldest dead containers by enforcing MaxContainers
+// * remove dead sandboxes with zero containers.
+func (m *kubeGenericRuntimeManager) GarbageCollect(gcPolicy kubecontainer.ContainerGCPolicy, allSourcesReady bool) error {
+	evictablePods, err := m.getEvictablePods(gcPolicy.MinAge)
+	if err != nil {
+		glog.Warningf("Get evictable pods failed: %v", err)
```

Comment:
`getEvictablePods` to ease debugging in the future instead of `Get evictable pods`